> `error: undefined reference to '__fpclassifyf'`, `floor`, `log10` etc. Is `libm` not being linked?

I'm struggling with the mixture of (UT_)-DRUNTIMESO and DRUNTIMESOLIB...

--- a/.azure-pipelines/windows.sh
+++ b/.azure-pipelines/windows.sh
@@ -164,6 +164,9 @@ if [ "$MODEL" != "32omf" ] ; then
     # run some tests for shared druntime
     # the test_runner links against libdruntime-ut.dll and runs all unittests
     #  no matter what module name is passed in
+
+    # no separate output for static or shared builds, so clean directories to force rebuild
+    rm -rf test/shared/generated
     # running additional tests needs druntime_shared.dll on the path
     export PATH="$PATH:$PWD/generated/windows/release/$MODEL"
     "$GNU_MAKE" -j$N unittest MODEL=$MODEL SHARED=1 DMD="$DMD_BIN_PATH" CC="$CC" \

--- a/.azure-pipelines/windows.sh
+++ b/.azure-pipelines/windows.sh
@@ -160,6 +160,16 @@ fi
 cd "$DMD_DIR/druntime"
 "$GNU_MAKE" -j$N MODEL=$MODEL DMD="$DMD_BIN_PATH" CC="$CC" unittest
 
+if [ "$MODEL" != "32omf" ] ; then
+    # run some tests for shared druntime
+    # the test_runner links against libdruntime-ut.dll and runs all unittests
+    #  no matter what module name is passed in
+    # running additional tests needs druntime_shared.dll on the path
+    export PATH="$PATH:$PWD/generated/windows/release/$MODEL"
+    "$GNU_MAKE" -j$N unittest MODEL=$MODEL SHARED=1 DMD="$DMD_BIN_PATH" CC="$CC" \
+        UT_MODULES=../generated/windows/release/$MODEL/unittest/object ADDITIONAL_TESTS=test/shared
+fi
+
 ################################################################################
 # Build and run Phobos unittests
 ################################################################################

--- a/compiler/src/dmd/backend/backconfig.d
+++ b/compiler/src/dmd/backend/backconfig.d
@@ -141,10 +141,10 @@ extern (C) void out_config_init(
             cfg.objfmt = mscoff ? OBJ_MSCOFF : OBJ_OMF;
             if (mscoff)
                 cfg.flags |= CFGnoebp;    // test suite fails without this
-            if (dataimports)
-                cfg.flags2 |= CFG2noreadonly;
         }
 
+        if (dataimports)
+            cfg.flags2 |= CFG2noreadonly;
         if (exe)
             cfg.wflags |= WFexe;         // EXE file only optimizations
         cfg.flags4 |= CFG4underscore;

--- a/compiler/src/dmd/cli.d
+++ b/compiler/src/dmd/cli.d
@@ -340,9 +340,9 @@ dmd -cov -unittest myprog.d
             "Windows only: select symbols to dllimport (none/defaultLibsOnly/all)",
             `Which global variables to dllimport implicitly if not defined in a root module
             $(UL
-                $(LI $(I none): None
-                $(LI $(I defaultLibsOnly): Only druntime/Phobos symbols
-                $(LI $(I all): All
+                $(LI $(I none): None)
+                $(LI $(I defaultLibsOnly): Only druntime/Phobos symbols)
+                $(LI $(I all): All)
             )`,
         ),
         Option("extern-std=<standard>",

--- a/compiler/src/dmd/e2ir.d
+++ b/compiler/src/dmd/e2ir.d
@@ -57,6 +57,7 @@ import dmd.toctype;
 import dmd.toir;
 import dmd.tokens;
 import dmd.toobj;
+import dmd.typinf;
 import dmd.visitor;
 
 import dmd.backend.cc;
@@ -508,17 +509,32 @@ bool isDllImported(Dsymbol var)
     if (driverParams.symImport == SymImport.none)
         return false;
     if (auto vd = var.isDeclaration())
-        if (!vd.isStatic() || vd.isThreadlocal())
+        if (!vd.isDataseg() || vd.isThreadlocal())
             return false;
-    if (driverParams.symImport != SymImport.all)
-        return false;  // todo: check std/core
-    auto m = var.getModule();
-    if (m)
-        return !m.isRoot();
+    if (var.isFuncDeclaration())
+        return false; // can always jump through import table
+    if (auto tid = var.isTypeInfoDeclaration())
+        if (builtinTypeInfo(tid.tinfo))
+            return true;
+    if (driverParams.symImport == SymImport.defaultLibsOnly)
+    {
+        auto m = var.getModule();
+        if (!m || !m.md)
+            return false;
+        const id = m.md.packages.length ? m.md.packages[0] : null;
+        if (id && id != Id.core && id != Id.std)
+            return false;
+        else if (!id && m.md.id != Id.std && m.md.id != Id.object)
+            return false;
+    }
+    else if (driverParams.symImport != SymImport.all)
+        return false;
+    if (auto mod = var.isModule())
+        return !mod.isRoot(); // non-root ModuleInfo symbol
     if (var.inNonRoot())
         return true; // not instantiated, and defined in non-root
-    if (var.isInstantiated()) // && !defineOnDeclare(sym, false))
-        return true; // instantiated but potentially culled (needsCodegen())
+    if (auto ti = var.isInstantiated()) // && !defineOnDeclare(sym, false))
+        return !ti.needsCodegen(); // instantiated but potentially culled (needsCodegen())
     if (auto vd = var.isVarDeclaration())
         if (vd.storage_class & STC.extern_)
             return true; // externally defined global variable

--- a/druntime/Makefile
+++ b/druntime/Makefile
@@ -483,7 +483,7 @@ $(UT_DRUNTIME): $(OBJS) $(SRCS) $(DMD)
 	$(DMD) $(UDFLAGS) -shared $(UTFLAGS) -of$@ $(SRCS) $(OBJS) $(LINKDL) -defaultlib= $(if $(findstring $(OS),windows),user32.lib -L/IMPLIB:$(UT_DRUNTIMELIB),) $(SOLIBS)
 
 $(ROOT)/unittest/test_runner$(DOTEXE): $(UT_DRUNTIME) src/test_runner.d $(DMD)
-	$(DMD) $(UDFLAGS) -of$@ src/test_runner.d -L$(UT_DRUNTIMELIB) -defaultlib= $(if $(findstring $(OS),windows),user32.lib,-L-lpthread -L-lm)
+	$(DMD) $(UDFLAGS) -of$@ src/test_runner.d -L$(UT_DRUNTIMELIB) -defaultlib= $(if $(findstring $(OS),windows),-dllimport=defaultLibsOnly user32.lib,-L-lpthread -L-lm)
 
 endif
 

--- a/druntime/test/common.mak
+++ b/druntime/test/common.mak
@@ -19,6 +19,7 @@ LDL:=$(subst -L,,$(LINKDL)) # -ldl
 SRC:=src
 GENERATED:=./generated
 ROOT:=$(GENERATED)/$(OS)/$(BUILD)/$(MODEL)
+DRUNTIME_IMPLIB:=$(subst .dll,.lib,$(DRUNTIMESO))
 
 MODEL_FLAG:=$(if $(findstring $(MODEL),default),,-m$(MODEL))
 CFLAGS_BASE:=$(if $(findstring $(OS),windows),/Wall,$(MODEL_FLAG) $(PIC) -Wall)
@@ -27,7 +28,7 @@ ifeq (osx64,$(OS)$(MODEL))
 endif
 DFLAGS:=$(MODEL_FLAG) $(PIC) -w -I../../src -I../../import -I$(SRC) -defaultlib= -preview=dip1000 $(if $(findstring $(OS),windows),,-L-lpthread -L-lm $(LINKDL))
 # LINK_SHARED may be set by importing makefile
-DFLAGS+=$(if $(LINK_SHARED),-L$(DRUNTIMESO) $(if $(findstring $(OS),windows),-dllimport=defaultLibsOnly),-L$(DRUNTIME))
+DFLAGS+=$(if $(LINK_SHARED),-L$(DRUNTIME_IMPLIB) $(if $(findstring $(OS),windows),-dllimport=defaultLibsOnly),-L$(DRUNTIME))
 ifeq ($(BUILD),debug)
     DFLAGS+=-g -debug
     CFLAGS:=$(CFLAGS_BASE) $(if $(findstring $(OS),windows),/Zi,-g)

--- a/druntime/test/shared/Makefile
+++ b/druntime/test/shared/Makefile
@@ -1,5 +1,5 @@
 # druntime DLL not supported on Windows yet
-LINK_SHARED:=$(if $(findstring $(OS),windows),,1)
+LINK_SHARED:=$(if $(findstring $(OS),windows),$(SHARED),1)
 
 include ../common.mak # affected by LINK_SHARED!
 

